homeassistant/components/imap/coordinator.py

```diff
@@ -23,7 +24,11 @@ async def connect_to_server(data: dict[str, Any]) -> IMAP4_SSL:
     client = IMAP4_SSL(data[CONF_SERVER], data[CONF_PORT])
     await client.wait_hello_from_server()
     await client.login(data[CONF_USERNAME], data[CONF_PASSWORD])
+    if client.protocol.state != AUTH:
+        raise InvalidAuth
     await client.select(data[CONF_FOLDER])
+    if client.protocol.state != SELECTED:
+        raise InvalidFolder
```

Comment:
Further testing showed that invalid authentication or wrong folder selection doesn't raise an exception by the library but rather update the state of the client object. So I added the code to raise some exceptions based on the detected state.